File: objc/Source/DIDComm-gRPC/DIDComm.m
//
//  DIDComm.m
//  DIDComm-gRPC
//
//  Created by Tomislav Markovski on 11/10/20.
//

#import "DIDComm.h"
#import "NativeFunc.h"
#import <Protobuf/GPBMessage.h>

@implementation DIDComm


+ (SignResponse *)sign:(SignRequest *)request
             withError:(NSError *_Nullable*_Nullable)error
{
    NSData* requestData = request.data;
    
    ByteBuffer req;
    req.len = requestData.length;
    req.data = (uint8_t *)[requestData bytes];

    ByteBuffer *response = (ByteBuffer*) malloc(sizeof(ByteBuffer));
    ExternError *err = (ExternError*) malloc(sizeof(ExternError));

    if (didcomm_sign(req, response, err) != 0) {
        *error = [self errorFromExternError:err];
        return nil;
    }
    
    NSData* responseData = [[NSData alloc] initWithBytesNoCopy:response->data length:(NSUInteger)response->len freeWhenDone:true];
    
    return [SignResponse parseFromData:responseData error:error];
}

+ (VerifyResponse *)verify:(VerifyRequest *)request
                 withError:(NSError *_Nullable*_Nullable)error
{
    NSData* requestData = request.data;
    
    ByteBuffer req;
    req.len = requestData.length;
    req.data = (uint8_t *)[requestData bytes];

    ByteBuffer *response = (ByteBuffer*) malloc(sizeof(ByteBuffer));
    ExternError *err = (ExternError*) malloc(sizeof(ExternError));

    if (didcomm_verify(req, response, err) != 0) {
        *error = [self errorFromExternError:err];
        return nil;
    }
    
    NSData* responseData = [[NSData alloc] initWithBytesNoCopy:response->data length:(NSUInteger)response->len freeWhenDone:true];
    
    return [VerifyResponse parseFromData:responseData error:error];
}

+ (PackResponse *)pack:(PackRequest *)request
             withError:(NSError *_Nullable*_Nullable)error
{
    NSData* requestData = request.data;
    
    ByteBuffer req;
    req.len = requestData.length;
    req.data = (uint8_t *)[requestData bytes];

    ByteBuffer *response = (ByteBuffer*) malloc(sizeof(ByteBuffer));
    ExternError *err = (ExternError*) malloc(sizeof(ExternError));

    if (didcomm_pack(req, response, err) != 0) {
        *error = [self errorFromExternError:err];
    }
    
    NSData* responseData = [[NSData alloc] initWithBytesNoCopy:response->data length:(NSUInteger)response->len freeWhenDone:true];
    
    return [PackResponse parseFromData:responseData error:error];
}

+ (UnpackResponse *)unpack:(UnpackRequest *)request
                 withError:(NSError *_Nullable*_Nullable)error
{
    NSData* requestData = request.data;
    
    ByteBuffer req;
    req.len = requestData.length;
    req.data = (uint8_t *)[requestData bytes];

    ByteBuffer *response = (ByteBuffer*) malloc(sizeof(ByteBuffer));
    ExternError *err = (ExternError*) malloc(sizeof(ExternError));

    if (didcomm_unpack(req, response, err) != 0) {
        @throw [self errorFromExternError:err];
    }
    
    NSData* responseData = [[NSData alloc] initWithBytesNoCopy:response->data length:(NSUInteger)response->len freeWhenDone:true];
    
    return [UnpackResponse parseFromData:responseData error:error];
}


static NSString *const DMErrorDomain = @"DIDCommError";

+ (NSError *)errorFromExternError:(ExternError *)error {

    NSMutableDictionary *userInfo = [NSMutableDictionary new];

    if (error->message != NULL) {
        [userInfo setValue:[NSString stringWithUTF8String:error->message] forKey:@"message"];
        free(error->message);
    }

    NSError *result = [NSError errorWithDomain:DMErrorDomain code:error->code userInfo:userInfo];

    free(error);

    return result;
}

@end

File: java/src/main/java/trinsic/okapi/DidComm.java
package trinsic.okapi;

import com.google.protobuf.InvalidProtocolBufferException;
import trinsic.okapi.transport.v1.Transport;

public class DidComm extends OkapiNative {
    public static Transport.PackResponse pack(Transport.PackRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        var result = getNativeLibrary().didcomm_pack(requestBuffer, responseBuffer, errBuffer);
        errBuffer.raiseError(result);
        return Transport.PackResponse.parseFrom(bufferToByteArray(responseBuffer));
    }

    public static Transport.UnpackResponse unpack(Transport.UnpackRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        var result = getNativeLibrary().didcomm_unpack(requestBuffer, responseBuffer, errBuffer);
        errBuffer.raiseError(result);
        return Transport.UnpackResponse.parseFrom(bufferToByteArray(responseBuffer));
    }

    public static Transport.SignResponse sign(Transport.SignRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        var result = getNativeLibrary().didcomm_sign(requestBuffer, responseBuffer, errBuffer);
        errBuffer.raiseError(result);
        return Transport.SignResponse.parseFrom(bufferToByteArray(responseBuffer));
    }

    public static Transport.VerifyResponse verify(Transport.VerifyRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        var result = getNativeLibrary().didcomm_verify(requestBuffer, responseBuffer, errBuffer);
        errBuffer.raiseError(result);
        return Transport.VerifyResponse.parseFrom(bufferToByteArray(responseBuffer));
    }
}

